File: webflux-dialogue/src/main/java/com/study/webflux/rag/domain/monitoring/model/UsageAnalytics.java
package com.study.webflux.rag.domain.monitoring.model;

import java.time.Instant;
import java.util.List;

import com.study.webflux.rag.domain.cost.model.CostInfo;

/** 파이프라인 실행의 종합 분석 데이터를 표현합니다. */
public record UsageAnalytics(
	String pipelineId,
	String status,
	Instant timestamp,
	UserRequest userRequest,
	LlmUsage llmUsage,
	RetrievalMetrics retrievalMetrics,
	TtsMetrics ttsMetrics,
	ResponseMetrics responseMetrics,
	CostInfo costInfo
) {
	/** 사용자 입력 관련 지표입니다. */
	public record UserRequest(
		String inputText,
		int inputLength,
		String inputPreview) {
	}

	/** LLM 사용량과 생성 문장 지표입니다. */
	public record LlmUsage(
		String model,
		int promptTokens,
		int completionTokens,
		int totalTokens,
		List<String> generatedSentences,
		long completionTimeMillis) {
	}

	/** RAG 검색 관련 지표입니다. */
	public record RetrievalMetrics(
		int memoryCount,
		int documentCount,
		long retrievalTimeMillis) {
	}

	/** TTS 합성 관련 지표입니다. */
	public record TtsMetrics(
		int sentenceCount,
		int audioChunks,
		long synthesisTimeMillis,
		long audioLengthMillis) {
	}

	/** 응답 지연 및 실행 시간 지표입니다. */
	public record ResponseMetrics(
		long totalDurationMillis,
		Long firstResponseLatencyMillis,
		Long lastResponseLatencyMillis) {
	}

	public static Builder builder() {
		return new Builder();
	}

	public static class Builder {
		private String pipelineId;
		private String status;
		private Instant timestamp;
		private UserRequest userRequest;
		private LlmUsage llmUsage;
		private RetrievalMetrics retrievalMetrics;
		private TtsMetrics ttsMetrics;
		private ResponseMetrics responseMetrics;
		private CostInfo costInfo;

		public Builder pipelineId(String pipelineId) {
			this.pipelineId = pipelineId;
			return this;
		}

		public Builder status(String status) {
			this.status = status;
			return this;
		}

		public Builder timestamp(Instant timestamp) {
			this.timestamp = timestamp;
			return this;
		}

		public Builder userRequest(UserRequest userRequest) {
			this.userRequest = userRequest;
			return this;
		}

		public Builder llmUsage(LlmUsage llmUsage) {
			this.llmUsage = llmUsage;
			return this;
		}

		public Builder retrievalMetrics(RetrievalMetrics retrievalMetrics) {
			this.retrievalMetrics = retrievalMetrics;
			return this;
		}

		public Builder ttsMetrics(TtsMetrics ttsMetrics) {
			this.ttsMetrics = ttsMetrics;
			return this;
		}

		public Builder responseMetrics(ResponseMetrics responseMetrics) {
			this.responseMetrics = responseMetrics;
			return this;
		}

		public Builder costInfo(CostInfo costInfo) {
			this.costInfo = costInfo;
			return this;
		}

		public UsageAnalytics build() {
			return new UsageAnalytics(
				pipelineId,
				status,
				timestamp,
				userRequest,
				llmUsage,
				retrievalMetrics,
				ttsMetrics,
				responseMetrics,
				costInfo);
		}
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/monitoring/model/UsageAnalytics.kt
package com.miyou.app.domain.monitoring.model

import com.miyou.app.domain.cost.model.CostInfo
import java.time.Instant

/**
 * 대화 파이프라인 실행의 종합 분석 데이터를 표현합니다.
 */
data class UsageAnalytics(
    val pipelineId: String,
    val status: String,
    val timestamp: Instant?,
    val userRequest: UserRequest?,
    val llmUsage: LlmUsage?,
    val retrievalMetrics: RetrievalMetrics?,
    val ttsMetrics: TtsMetrics?,
    val responseMetrics: ResponseMetrics?,
    val costInfo: CostInfo?,
) {
    /**
     * 사용자 입력 관련 지표입니다.
     */
    data class UserRequest(
        val inputText: String,
        val inputLength: Int,
        val inputPreview: String?,
    )

    /**
     * LLM 사용량과 생성 문장 지표입니다.
     */
    data class LlmUsage(
        val model: String,
        val promptTokens: Int,
        val completionTokens: Int,
        val totalTokens: Int,
        val generatedSentences: List<String>,
        val completionTimeMillis: Long,
    )

    /**
     * RAG 검색 관련 지표입니다.
     */
    data class RetrievalMetrics(
        val memoryCount: Int,
        val documentCount: Int,
        val retrievalTimeMillis: Long,
    )

    /**
     * TTS 합성 관련 지표입니다.
     */
    data class TtsMetrics(
        val sentenceCount: Int,
        val audioChunks: Int,
        val synthesisTimeMillis: Long,
        val audioLengthMillis: Long,
    )

    /**
     * 응답 지연 및 실행 시간 지표입니다.
     */
    data class ResponseMetrics(
        val totalDurationMillis: Long,
        val firstResponseLatencyMillis: Long?,
        val lastResponseLatencyMillis: Long?,
    )

    class Builder {
        private var pipelineId: String? = null
        private var status: String? = null
        private var timestamp: Instant? = null
        private var userRequest: UserRequest? = null
        private var llmUsage: LlmUsage? = null
        private var retrievalMetrics: RetrievalMetrics? = null
        private var ttsMetrics: TtsMetrics? = null
        private var responseMetrics: ResponseMetrics? = null
        private var costInfo: CostInfo? = null

        fun pipelineId(pipelineId: String?): Builder = apply { this.pipelineId = pipelineId }

        fun status(status: String?): Builder = apply { this.status = status }

        fun timestamp(timestamp: Instant?): Builder = apply { this.timestamp = timestamp }

        fun userRequest(userRequest: UserRequest?): Builder = apply { this.userRequest = userRequest }

        fun llmUsage(llmUsage: LlmUsage?): Builder = apply { this.llmUsage = llmUsage }

        fun retrievalMetrics(retrievalMetrics: RetrievalMetrics?): Builder =
            apply {
                this.retrievalMetrics =
                    retrievalMetrics
            }

        fun ttsMetrics(ttsMetrics: TtsMetrics?): Builder = apply { this.ttsMetrics = ttsMetrics }

        fun responseMetrics(responseMetrics: ResponseMetrics?): Builder =
            apply { this.responseMetrics = responseMetrics }

        fun costInfo(costInfo: CostInfo?): Builder = apply { this.costInfo = costInfo }

        fun build(): UsageAnalytics =
            UsageAnalytics(
                pipelineId = pipelineId.orEmpty(),
                status = status.orEmpty(),
                timestamp = timestamp,
                userRequest = userRequest,
                llmUsage = llmUsage,
                retrievalMetrics = retrievalMetrics,
                ttsMetrics = ttsMetrics,
                responseMetrics = responseMetrics,
                costInfo = costInfo,
            )
    }

    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }
}
